cryptol-remote-api/python/cryptol/synchronous.py

```diff
@@ -0,0 +1,283 @@
+    connection : CryptolConnection
+
+    def __init__(self, connection : CryptolConnection):
+        self.connection = connection
+
+    def load_file(self, filename : str) -> None:
+        """Load a filename as a Cryptol module, like ``:load`` at the Cryptol
+        REPL.
+        """
+        self.connection.load_file(filename).result()
+
+    def load_module(self, module_name : str) -> None:
+        """Load a Cryptol module, like ``:module`` at the Cryptol REPL."""
+        self.connection.load_module(module_name).result()
+
+    def extend_search_path(self, *dir : str) -> None:
+        """Extend the search path for loading Cryptol modules."""
+        self.connection.extend_search_path(*dir).result()
+
+    def eval(self, expression : Any) -> CryptolValue:
+        """Evaluate a Cryptol expression, with the result represented
+        according to :ref:`cryptol-json-expression`, with Python datatypes
+        standing for their JSON equivalents.
+        """
+        return from_cryptol_arg(self.connection.eval_raw(expression).result())
+
+    def call(self, fun : str, *args : List[Any]) -> CryptolValue:
+        """Evaluate a Cryptol functiom by name, with the arguments and the
+        result represented according to :ref:`cryptol-json-expression`, with
+        Python datatypes standing for their JSON equivalents.
+        """
+        return from_cryptol_arg(self.connection.call_raw(fun, *args).result())
+
+    def check(self, expr : Any, *, num_tests : Union[Literal['all'], int, None] = None) -> CheckReport:
+        """Tests the validity of a Cryptol expression with random inputs. The expression must be a function with
+        return type ``Bit``.
+        If ``num_tests`` is ``"all"`` then the expression is tested exhaustively (i.e., against all possible inputs).
+        If ``num_tests`` is omitted, Cryptol defaults to running 100 tests.
+        """
+        return to_check_report(self.connection.check_raw(expr, num_tests=num_tests).result())
+
+    def check_type(self, code : Any) -> cryptoltypes.CryptolType:
+        """Check the type of a Cryptol expression, represented according to
+        :ref:`cryptol-json-expression`, with Python datatypes standing for
+        their JSON equivalents.
+        """
+        return cryptoltypes.to_type(self.connection.check_type(code).result()['type'])
+
+    @overload
+    def sat(self, expr : Any, solver : OnlineSolver = Z3, count : int = 1) -> Union[Satisfiable, Unsatisfiable]: ...
+    @overload
+    def sat(self, expr : Any, solver : OfflineSolver, count : int = 1) -> OfflineSmtQuery: ...
+
+    def sat(self, expr : Any, solver : Solver = Z3, count : int = 1) -> Union[Satisfiable, Unsatisfiable, OfflineSmtQuery]:
+        """Check the satisfiability of a Cryptol expression, represented according to
+        :ref:`cryptol-json-expression`, with Python datatypes standing for
+        their JSON equivalents. Use the solver named `solver`, and return up to
+        `count` solutions.
```

Comment:
We should add a note in this docstring regarding being able to boolean test `Satisfiable` and `Unsatisfiable` (i.e., that the latter is falsy).